README.md
# faasd CDK

![NPN version](https://img.shields.io/npm/v/@strongishllama/faasd-cdk)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://raw.githubusercontent.com/strongishllama/faasd-cdk/main/LICENSE)
![GitHub Workflow Status](https://img.shields.io/github/workflow/status/strongishllama/faasd-cdk/Release)

This project contains CDK constructs to deploy the [faasd](https://github.com/openfaas/faasd) on AWS. Currently, there is only a single construct to deploy faasd to an EC2 instance. Be warned, this construct is very much in its prototyping phase.

## Installation
Install this package with the following command.
```
npm install @strongishllama/faasd-cdk
```

Add the import at the top of your ```.ts``` file.
```ts
import * as faasd from '@strongishllama/faasd-cdk';
```

Then add the construct to your CDK stack, filling out the properties.
```ts
new faasd.Instance(this, 'faasd-instance', {
  region: 'ap-southeast-2',
  baseDomainName: 'example.com',
  fullDomainName: 'faasd.example.com',
  emailAddress: 'webmaster@example.com',
  vpc: ec2.Vpc.fromLookup(this, 'vpc', {
    vpcId: 'vpc-123456'
  }),
  amiId: 'ami-123456'
});
```

## Testing
There are two types of tests in this project, CDK tests and deployment tests.

CDK tests test the CloudFormation output by the constructs, these can be run with the following command.
```
npm run test
```

Deployment tests import the constructs so they can be manually tested before making a package release. The plan is to automate these in the future.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.2.0 - 2021-10-30
### Added
* **BREAKING**: Added ```InstanceProps.amiId``` property.
* Added the ability to specify a custom AMI ID on the ```Instance``` stack.

### Removed
* **BREAKING**: Removed unneeded ```InstanceProps.account``` property.

## v0.1.0 - 2021-10-10
### Added
* Added the ```Instance``` construct.
* Added tests for the ```Instance``` construct.
* Added a license.
* Added a changelog.
* Added a code of conduct.
* Added contributing guidelines.
* Added a pull request template.
* Added issue templates.


test/cdk/faasd-cdk.test.ts
import * as assert from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as faasd from '../../lib/index';

test('Instance', () => {
  const app = new cdk.App();
  const env = {
    account: '0123456789',
    region: 'ap-southeast-2'
  }
  const stack = new cdk.Stack(app, 'test-stack', {
    env: env
  });

  // WHEN
  const props: faasd.InstanceProps = {
    region: env.region,
    baseDomainName: 'example.com',
    fullDomainName: 'faasd.example.com',
    emailAddress: 'webmaster@example.com',
    vpc: ec2.Vpc.fromLookup(stack, 'vpc', {
      vpcId: 'vpc-123456'
    }),
    amiId: 'ami-0567f647e75c7bc05'
  };
  new faasd.Instance(stack, 'instance', props);

  // THEN
  // Security group.
  assert.expect(stack).to(assert.countResourcesLike('AWS::EC2::SecurityGroup', 1, {
    'SecurityGroupEgress': [
      {
        'CidrIp': '0.0.0.0/0',
        'IpProtocol': '-1'
      }
    ],
    'SecurityGroupIngress': [
      {
        'CidrIp': '0.0.0.0/0',
        'FromPort': 80,
        'IpProtocol': 'tcp',
        'ToPort': 80
      },
      {
        'CidrIpv6': '::/0',
        'FromPort': 80,
        'IpProtocol': 'tcp',
        'ToPort': 80
      },
      {
        'CidrIp': '0.0.0.0/0',
        'FromPort': 443,
        'IpProtocol': 'tcp',
        'ToPort': 443
      },
      {
        'CidrIpv6': '::/0',
        'FromPort': 443,
        'IpProtocol': 'tcp',
        'ToPort': 443
      }
    ],
    'VpcId': props.vpc.vpcId
  }));
  // Role.
  assert.expect(stack).to(assert.countResourcesLike('AWS::IAM::Role', 1, {
    'AssumeRolePolicyDocument': {
      'Statement': [
        {
          'Action': 'sts:AssumeRole',
          'Effect': 'Allow',
          'Principal': {
            'Service': 'ec2.amazonaws.com'
          }
        }
      ]
    },
    'ManagedPolicyArns': [
      'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore'
    ]
  }));
  // Policy.
  assert.expect(stack).to(assert.countResourcesLike('AWS::IAM::Policy', 1, {
    'PolicyDocument': {
      'Statement': [
        {
          'Action': [
            'secretsmanager:GetSecretValue',
            'secretsmanager:DescribeSecret'
          ],
          'Effect': 'Allow',
          'Resource': {
            'Ref': assert.anything()
          }
        }
      ],
      'Version': '2012-10-17'
    },
    'Roles': [
      {
        'Ref': assert.anything()
      }
    ]
  }));
  // Instance.
  assert.expect(stack).to(assert.countResourcesLike('AWS::EC2::Instance', 1, {
    'ImageId': props.amiId,
    'InstanceType': `${props.instanceClass ?? 't3a'}.${props.instanceSize ?? 'micro'}`,
    'SecurityGroupIds': [
      {
        'Fn::GetAtt': [
          assert.anything(),
          'GroupId'
        ]
      }
    ],
  }));
  // Elastic IP.
  assert.expect(stack).to(assert.countResources('AWS::EC2::EIP', 1));
  assert.expect(stack).to(assert.countResources('AWS::EC2::EIPAssociation', 1));
  // Record set.
  assert.expect(stack).to(assert.countResourcesLike('AWS::Route53::RecordSet', 1, {
    'Name': `${props.fullDomainName ?? props.baseDomainName}.`,
    'Type': 'A',
    'HostedZoneId': assert.anything(),
    'ResourceRecords': [
      {
        'Ref': assert.anything()
      }
    ],
    'TTL': '1800'
  }));
  assert.expect(stack).to(assert.countResources('AWS::EC2::EIPAssociation', 1));
  // Record set.
  assert.expect(stack).to(assert.countResourcesLike('AWS::Route53::RecordSet', 1, {
    'Name': `${props.fullDomainName ?? props.baseDomainName}.`,
    'Type': 'A',
    'HostedZoneId': assert.anything(),
    'ResourceRecords': [
      {
        'Ref': assert.anything()
      }
    ],
    'TTL': '1800'
  }));
  // Secret.
  assert.expect(stack).to(assert.countResourcesLike('AWS::SecretsManager::Secret', 1, {
    'GenerateSecretString': {}
  }));
});


test/deploy/lib/deploy-stack.ts
import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as faasd from '../../../lib';

export interface DeployStackProps extends cdk.StackProps {
  readonly baseDomainName: string;
  readonly fullDomainName?: string;
  readonly emailAddress: string;
  readonly vpcId: string;
}

export class DeployStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props: DeployStackProps) {
    super(scope, id, props);

    if (props.env === undefined || props.env.account === undefined || props.env.region === undefined) {
      throw new Error('DeployStackProps.env property must be fully defined');
    }

    new faasd.Instance(this, 'faasd', {
      region: props.env.region,
      baseDomainName: props.baseDomainName,
      fullDomainName: props.fullDomainName,
      emailAddress: props.emailAddress,
      vpc: new ec2.Vpc(this, 'vpc', {
        cidr: '10.0.0.0/16',
        subnetConfiguration: [
          {
            cidrMask: 24,
            name: 'ingress',
            subnetType: ec2.SubnetType.PUBLIC
          }
        ]
      }),
      amiId: 'ami-0567f647e75c7bc05'
    });
  }
}


lib/index.ts
import * as fs from 'fs';
import * as path from 'path';
import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';
import * as route53 from '@aws-cdk/aws-route53';
import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
import * as handlebars from 'handlebars';

export interface InstanceProps {
  /**
   * The region being deployed into.
   */
  readonly region: string;
  /**
   * The base domain to use. This domain must have a hosted zone in Route53.
   */
  readonly baseDomainName: string;
  /**
   * The full domain name to use. If left undefined the base domain name
   * will be used.
   *
   * @default baseDomainName
   */
  readonly fullDomainName?: string;
  /**
   * The email to register with LetsEncrypt.
   */
  readonly emailAddress: string;
  /**
   * The VPC for the instance to be deploy into.
   */
  readonly vpc: ec2.IVpc;
  /**
   * The class of the EC2 instance being deployed.
   *
   * @default ec2.InstanceClass.T3A
   */
  readonly instanceClass?: ec2.InstanceClass;
  /**
   * The size of the EC2 instance being deployed.
   *
   * @default ec2.InstanceSize.MICRO
   */
  readonly instanceSize?: ec2.InstanceSize;
  /**
   * The ID of the AMI to use when deploying your instance.
   */
  readonly amiId: string;
}

export class Instance extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: InstanceProps) {
    super(scope, id);

    // Create the security group.
    const securityGroup = new ec2.SecurityGroup(this, 'security-group', {
      vpc: props.vpc
    });
    securityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(80));
    securityGroup.addIngressRule(ec2.Peer.anyIpv6(), ec2.Port.tcp(80));
    securityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(443));
    securityGroup.addIngressRule(ec2.Peer.anyIpv6(), ec2.Port.tcp(443));

    // Create the instance.
    const instance = new ec2.Instance(this, 'instance', {
      vpc: props.vpc,
      securityGroup: securityGroup,
      instanceType: ec2.InstanceType.of(props.instanceClass ?? ec2.InstanceClass.T3A, props.instanceSize ?? ec2.InstanceSize.MICRO),
      machineImage: ec2.MachineImage.genericLinux({ [props.region] : props.amiId }),
      role: new iam.Role(this, 'role', {
        assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
        managedPolicies: [
          iam.ManagedPolicy.fromManagedPolicyArn(this, 'ssm-policy', 'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore')
        ]
      })
    });

    // Create an elastic IP and associate it with the instance.
    const eip = new ec2.CfnEIP(this, 'eip');
    new ec2.CfnEIPAssociation(this, 'eip-association', {
      eip: eip.ref,
      instanceId: instance.instanceId
    });

    // Fetch the hosted zone.
    const hostedZone = route53.HostedZone.fromLookup(this, 'hosted-zone', {
      domainName: props.baseDomainName
    });

    // Create an A record for the full domain name.
    new route53.ARecord(this, 'a-record', {
      zone: hostedZone,
      recordName: props.fullDomainName,
      target: route53.RecordTarget.fromIpAddresses(eip.ref)
    });

    // Create the generated password and give the instance permission to read it.
    const passwordSecret = new secretsmanager.Secret(this, 'secret', {
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });
    passwordSecret.grantRead(instance);

    // Add the instance user data.
    instance.addUserData(...[
      '#!/bin/bash',
      // Piping logs.
      'exec > >(tee /var/log/user-data.log|logger -t user-data -s 2>/dev/console) 2>&1',
      // Installing updates.
      'apt update',
      'apt upgrade -y',
      'apt autoclean',
      'apt autoremove -y',
      // Installing dependencies.
      'apt install runc bridge-utils make unzip debian-keyring debian-archive-keyring apt-transport-https -y',
      // Installing the AWS CLI.
      'curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "aws-cli-v2.zip"',
      'unzip aws-cli-v2.zip',
      './aws/install',
      'rm aws-cli-v2.zip',
      'rm -rf aws',
      // Installing Containerd.
      'sudo apt install containerd -y',
      '/sbin/sysctl -w net.ipv4.conf.all.forwarding=1',
      // Installing CNI.
      'mkdir -p /opt/cni/bin',
      'curl -sSL https://github.com/containernetworking/plugins/releases/download/v0.8.5/cni-plugins-linux-amd64-v0.8.5.tgz | tar -xz -C /opt/cni/bin',
      // Installing OpenFaas.
      'mkdir -p /go/src/github.com/openfaas',
      'mkdir -p /var/lib/faasd/secrets',
      `echo $(aws secretsmanager get-secret-value --secret-id ${passwordSecret.secretName} --query SecretString --output text) > /var/lib/faasd/secrets/basic-auth-password`,
      'echo admin > /var/lib/faasd/secrets/basic-auth-user',
      // Installing Faasd.
      'cd /go/src/github.com/openfaas',
      'git clone --depth 1 --branch 0.13.0 https://github.com/openfaas/faasd',
      'curl -fSLs "https://github.com/openfaas/faasd/releases/download/0.13.0/faasd" --output "/usr/local/bin/faasd"',
      'chmod a+x "/usr/local/bin/faasd"',
      'cd /go/src/github.com/openfaas/faasd',
      '/usr/local/bin/faasd install',
      // Installing Caddy.
      'mkdir -p /etc/caddy',
      `curl -1sLf 'https://dl.cloudsmith.io/public/caddy/stable/gpg.key' | sudo tee /etc/apt/trusted.gpg.d/caddy-stable.asc`,
      `curl -1sLf 'https://dl.cloudsmith.io/public/caddy/stable/debian.deb.txt' | sudo tee /etc/apt/sources.list.d/caddy-stable.list`,
      'apt update',
      'apt install caddy -y',
      `echo "${handlebars.compile(fs.readFileSync(path.join(__dirname, 'assets/Caddyfile')).toString())(props)}" > /etc/caddy/Caddyfile`,
      'cd /etc/caddy',
      'caddy reload'
    ]);

    // Output the important values.
    new cdk.CfnOutput(this, 'a-record-url', {
      value: `https://${props.fullDomainName ?? props.baseDomainName}`
    });
    new cdk.CfnOutput(this, 'username', {
      value: 'admin'
    });
    new cdk.CfnOutput(this, 'password-secret-url', {
      value: `https://${props.region}.console.aws.amazon.com/secretsmanager/home?region=${props.region}#!/secret?name=${passwordSecret.secretName}`
    });
  }
}
